### PR TITLE
Stop overanalyzing transitions from `repeated_receive`

### LINKED ISSUE #68: Stop overanalyzing transitions from `repeated_receive`
There are several places in bisimilarity proofs where we use `transition_from_repeated_receive` to learn exactly how a certain transition is structured, but actually only need to know that its target process family has the structure `T ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n`, because we can use the transition verbatim to construct a simulating transition and only need to know about the presence of `A ▹⇧∞ x. 𝒫 x` in a context, in order to prove that the target process families are in the respective relation up to some respectful function. We shall remove such current superfluous analysis of transitions and this way simplify the code considerably.

### PR DESCRIPTION
**Pull request by jeltsch**
This resolves #68.

### CODE REVIEW

**Comment by javierdiaz72** [comment]
https://github.com/input-output-hk/network-equivalences/blob/f21fa057b7db4404ba9c2c4cb9e1ef9eed1de43b/src/Thorn_Calculus-Core_Bisimilarities.thy#L1419

Shouldn't we also stop overanalyzing the transition in this case?

**Comment by jeltsch** [comment]
> https://github.com/input-output-hk/thorn-calculus/blob/f21fa057b7db4404ba9c2c4cb9e1ef9eed1de43b/src/Thorn_Calculus-Core_Bisimilarities.thy#L1419
> 
> Shouldn’t we also stop overanalyzing the transition in this case?

In fact, I started doing that, but then reverted to the original code. Why did I do so? Let me explain:

We definitely need to learn here that `α` has the form `IO η A n X`, since we have to use `parallel_left_io` two steps later; so obtaining `n` and `X` has to be done anyhow. Now, the current-iteration part of `S` can be described by solely referring to `𝒫` from the lemma statement and the newly obtained `n` and `X`: as `post_receive n X 𝒫`. It would simplify our `S`-term if we used just `T` for its current-iteration part, but this would require us to claim the existence of such a `T`. I reckoned that it would be less confusing to directly say what such a `T` would be, instead of mysteriously talking about a `T` that is actually not mysterious. :slightly_smiling_face: Similarly I opted for not claiming the existence of some `η`, but stated right away that it is `Receiving` (by using the notation `_ ▹ ⋆⇗_⇖ _`). Generally speaking, my approach was rather to reduce the number of unnecessary facts like `A' = A` instead of reducing term size.

Note how the situation in other places differs from this one: There you already know that `α` is something like `IO η A' n X`, which is enough for applying a lemma like `parallel_left_io` later. Our previous approach meant establishing what `η` and `A'` are and then writing down a concrete current-iteration term. In such a situation, it seems simpler to me to leave `η` and `A'` alone and just obtain a `T`.

**Comment by javierdiaz72** [comment]
> > https://github.com/input-output-hk/thorn-calculus/blob/f21fa057b7db4404ba9c2c4cb9e1ef9eed1de43b/src/Thorn_Calculus-Core_Bisimilarities.thy#L1419
> > 
> > Shouldn’t we also stop overanalyzing the transition in this case?
> 
> In fact, I started doing that, but then reverted to the original code. Why did I do so? Let me explain:
> 
> We definitely need to learn here that `α` has the form `IO η A n X`, since we have to use `parallel_left_io` two steps later; so obtaining `n` and `X` has to be done anyhow. Now, the current-iteration part of `S` can be described by solely referring to `𝒫` from the lemma statement and the newly obtained `n` and `X`: as `post_receive n X 𝒫`. It would simplify our `S`-term if we used just `T` for its current-iteration part, but this would require us to claim the existence of such a `T`. I reckoned that it would be less confusing to directly say what such a `T` would be, instead of mysteriously talking about a `T` that is actually not mysterious. slightly_smiling_face Similarly I opted for not claiming the existence of some `η`, but stated right away that it is `Receiving` (by using the notation `_ ▹ ⋆⇗_⇖ _`). Generally speaking, my approach was rather to reduce the number of unnecessary facts like `A' = A` instead of reducing term size.
> 
> Note how the situation in other places differs from this one: There you already know that `α` is something like `IO η A' n X`, which is enough for applying a lemma like `parallel_left_io` later. Our previous approach meant establishing what `η` and `A'` are and then writing down a concrete current-iteration term. In such a situation, it seems simpler to me to leave `η` and `A'` alone and just obtain a `T`.

That sounds sensible. :grin:

**Comment by jeltsch** [comment]
> https://github.com/input-output-hk/thorn-calculus/blob/f21fa057b7db4404ba9c2c4cb9e1ef9eed1de43b/src/Thorn_Calculus-Core_Bisimilarities.thy#L1419
> 
> Shouldn’t we also stop overanalyzing the transition in this case?

In fact, I started doing that, but then reverted to the original code. Why did I do so? Let me explain:

We definitely need to learn here that `α` has the form `IO η A n X`, since we have to use `parallel_left_io` two steps later; so obtaining `n` and `X` has to be done anyhow. Now, the current-iteration part of `S` can be described by solely referring to `𝒫` from the lemma statement and the newly obtained `n` and `X`: as `post_receive n X 𝒫`. It would simplify our `S`-term if we used just `T` for its current-iteration part, but this would require us to claim the existence of such a `T`. I reckoned that it would be less confusing to directly say what such a `T` would be, instead of mysteriously talking about a `T` that is actually not mysterious. :slightly_smiling_face: Similarly I opted for not claiming the existence of some `η`, but stated right away that it is `Receiving` (by using the notation `_ ▹ ⋆⇗_⇖ _`). Generally speaking, my approach was rather to reduce the number of unnecessary facts like `A' = A` instead of reducing term size.

Note how the situation in other places differs from this one: There you already know that `α` is something like `IO η A' n X`, which is enough for applying a lemma like `parallel_left_io` later. Our previous approach meant establishing what `η` and `A'` are and using a concrete current-iteration term. In such a situation, it seems simpler to me to leave `η` and `A'` alone and just obtain a `T`.

File: src/Thorn_Calculus-Core_Bisimilarities.thy [hunk 1]
```diff
@@ -1362,18 +1362,12 @@ text \<open>
 context begin
 
 private lemma post_left_receive_from_repeated_receive:
```
[the remark this answers is not in the document]
Comment (jeltsch) [reply src/Thorn_Calculus-Core_Bisimilarities.thy:1364]: Thanks for spotting this issue: indeed, this name doesn’t make sense anymore and neither does the one mentioned in the other comment below. Now I have to think of better names. :thinking:

File: src/Thorn_Calculus-Core_Bisimilarities.thy [hunk 2]
```diff
@@ -1362,18 +1362,12 @@ text \<open>
...
-  shows "
-    (post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n) \<parallel> Q \<guillemotleft> suffix n
-    \<sim>\<^sub>s
-    post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> Q) \<guillemotleft> suffix n"
+  shows "(T \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n) \<parallel> U \<guillemotleft> suffix n \<sim>\<^sub>s T \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> U) \<guillemotleft> suffix n"
   unfolding adapted_after_parallel
   using parallel_associativity .
 
 private lemma post_right_receive_from_repeated_receive:
```
[the remark this answers is not in the document]
Comment (jeltsch) [reply src/Thorn_Calculus-Core_Bisimilarities.thy:1369]: This lemma name even didn’t make sense before this change, but was being made senseless already when generalizing its statement to use `U`.

[the remark this answers is not in the document]
Comment (jeltsch) [reply src/Thorn_Calculus-Core_Bisimilarities.thy:1369]: Given that we have already generalized so much, we could go one step further and replace the occurrences of `A ▹⇧∞ x. 𝒫 x` by variables in the statements of both lemmas. Then some sensible names would come to my mind.

In line with these considerations, I propose to change these two lemma implementations to the following:
```isabelle
private lemma left_amended_double_suffix_adapted:
  shows "(R ∥ P « suffix n) ∥ Q « suffix n ∼⇩s R ∥ (P ∥ Q) « suffix n"
  unfolding adapted_after_parallel
  using parallel_associativity .

private lemma right_amended_double_suffix_adapted:
  shows "P « suffix n ∥ (R ∥ Q « suffix n) ∼⇩s R ∥ (P ∥ Q) « suffix n"
  unfolding adapted_after_parallel
  using parallel_left_commutativity .
```
What do you think?

[the remark this answers is not in the document]
Comment (javierdiaz72) [reply src/Thorn_Calculus-Core_Bisimilarities.thy:1369]: Looks good.

[the remark this answers is not in the document]
Comment (jeltsch) [reply src/Thorn_Calculus-Core_Bisimilarities.thy]: Implemented in 5f044b308165143d74f2d3385facb6ce6e73627c now. :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (javierdiaz72) [reply src/Thorn_Calculus-Core_Bisimilarities.thy]: Looks nice.  :slightly_smiling_face:

**Review by javierdiaz72 — APPROVED** [review]
Looks good to me.

**Pull request merged**